File: .pi/extensions/kamn-mvp/actor-receipts.ts
import type { Report } from "./evidence";

const RECEIPTS: ActorReceipt[] = [];
const TOOLS: ActorToolSpec[] = [
	{
		tool: "kamn_agent_a_register",
		agent: "agent_a",
		action: "register",
		scopeField: "agent_a_view_scope",
		artifactField: "agent_a_view_artifact",
		digestField: "agent_a_view_digest",
	},
	{
		tool: "kamn_agent_a_invoke_transaction",
		agent: "agent_a",
		action: "invoke_transaction",
		scopeField: "agent_a_view_scope",
		artifactField: "agent_a_view_artifact",
		digestField: "agent_a_view_digest",
	},
	{
		tool: "kamn_agent_b_register",
		agent: "agent_b",
		action: "register",
		scopeField: "agent_b_view_scope",
		artifactField: "agent_b_view_artifact",
		digestField: "agent_b_view_digest",
	},
	{
		tool: "kamn_agent_b_accept_task",
		agent: "agent_b",
		action: "accept_task",
		scopeField: "agent_b_view_scope",
		artifactField: "agent_b_view_artifact",
		digestField: "agent_b_view_digest",
	},
	{
		tool: "kamn_agent_c_verify_three_agent_proof",
		agent: "agent_c_verifier",
		action: "verify_proof",
		scopeField: "verifier_view_scope",
		artifactField: "agent_c_verifier_view_artifact",
		digestField: "agent_c_verifier_view_digest",
	},
];

export type ActorReceipt = {
	sequence: number;
	tool: string;
	agent: string;
	action: string;
	outcome: "PASS";
	report_path: string;
	view_scope: string;
	view_artifact: string;
	view_digest: string;
};

export function actorToolSpecs() {
	return TOOLS;
}

export function recordActorReceipt(reportPath: string, report: Report, tool: string) {
	const spec = requireToolSpec(tool);
	const claim = requireThreeAgentClaim(report);
	const sequence = receiptsForPath(reportPath).length + 1;
	const receipt = actorReceipt(reportPath, claim, spec, sequence);
	RECEIPTS.push(receipt);
	return receipt;
}

export function requiredActorReceipts(reportPath: string, report: Report) {
	if (!findClaim(report, "three_agent_escrow_verification")) return undefined;
	const receipts = receiptsForPath(reportPath);
	const missing = TOOLS.filter((tool) => !receipts.some((receipt) => receipt.tool === tool.tool));
	if (missing.length > 0) {
		throw new Error(`Missing KAMN actor tool receipts: ${missing.map((tool) => tool.tool).join(",")}`);
	}
	return receipts;
}

function actorReceipt(
	reportPath: string,
	claim: Record<string, unknown>,
	spec: ActorToolSpec,
	sequence: number,
): ActorReceipt {
	return {
		sequence,
		tool: spec.tool,
		agent: spec.agent,
		action: spec.action,
		outcome: "PASS",
		report_path: reportPath,
		view_scope: requireString(claim, spec.scopeField),
		view_artifact: requireString(claim, spec.artifactField),
		view_digest: requireString(claim, spec.digestField),
	};
}

function receiptsForPath(reportPath: string) {
	return RECEIPTS.filter((receipt) => receipt.report_path === reportPath);
}

function requireToolSpec(tool: string) {
	const spec = TOOLS.find((candidate) => candidate.tool === tool);
	if (!spec) throw new Error(`Unknown KAMN actor tool: ${tool}`);
	return spec;
}

function requireThreeAgentClaim(report: Report) {
	const claim = findClaim(report, "three_agent_escrow_verification");
	if (!claim) {
		throw new Error("KAMN report is missing three_agent_escrow_verification");
	}
	return claim;
}

function findClaim(report: Report, id: string) {
	return (report.claim_matrix ?? []).find((entry) => entry.id === id);
}

function requireString(claim: Record<string, unknown>, field: string): string {
	const value = claim[field];
	if (typeof value === "string") return value;
	throw new Error(`KAMN report three-agent claim is missing string field: ${field}`);
}

type ActorToolSpec = {
	tool: string;
	agent: string;
	action: string;
	scopeField: string;
	artifactField: string;
	digestField: string;
};

File: crates/kamn-e2e-harness/tests/mvp_demo_agent_harness_claim_contract/actor_receipts.rs
use std::path::Path;

pub(crate) fn valid_actor_receipts(root: &Path) -> String {
    format!(
        r#","three_agent_actor_tool_receipts":[{},{},{},{},{}]"#,
        receipt(root, 1, "kamn_agent_a_register", "agent_a", "register"),
        receipt(
            root,
            2,
            "kamn_agent_a_invoke_transaction",
            "agent_a",
            "invoke_transaction"
        ),
        receipt(root, 3, "kamn_agent_b_register", "agent_b", "register"),
        receipt(
            root,
            4,
            "kamn_agent_b_accept_task",
            "agent_b",
            "accept_task"
        ),
        receipt(
            root,
            5,
            "kamn_agent_c_verify_three_agent_proof",
            "agent_c_verifier",
            "verify_proof"
        )
    )
}

fn receipt(root: &Path, sequence: u64, tool: &str, agent: &str, action: &str) -> String {
    let view = view_for(root, agent);
    format!(
        r#"{{"sequence":{},"tool":"{}","agent":"{}","action":"{}","outcome":"PASS","report_path":"{}","view_scope":"{}","view_artifact":"{}","view_digest":"{}"}}"#,
        sequence,
        tool,
        agent,
        action,
        root.join("proof/report.json").display(),
        view.scope,
        view.artifact.display(),
        view.digest
    )
}

fn view_for(root: &Path, agent: &str) -> View {
    match agent {
        "agent_a" => View::participant(root, "agent-a-view.json", "agent-a-view-digest-7045"),
        "agent_b" => View::participant(root, "agent-b-view.json", "agent-b-view-digest-7045"),
        _ => View::verifier(root),
    }
}

struct View {
    scope: &'static str,
    artifact: std::path::PathBuf,
    digest: &'static str,
}

impl View {
    fn participant(root: &Path, file: &str, digest: &'static str) -> Self {
        Self {
            scope: "participant-private",
            artifact: root.join("proof").join(file),
            digest,
        }
    }

    fn verifier(root: &Path) -> Self {
        Self {
            scope: "restricted-public",
            artifact: root.join("proof/agent-c-verifier-view.json"),
            digest: "agent-c-view-digest-7045",
        }
    }
}
